### ISSUE TITLE
Crash with new Harmony from Workshop (ok with BepInEx)

### ISSUE THREAD

**Issue by Cilyena** [thomaswp/BeaverBuddies#25]
Update of : 27 juil. à 15h56
* Fixed crash on new game
* Removed Harmony dependency

Test :
With Harmony of last BepInEx, no bug
With Harmony of Workshop (remove BepInEx), bug at start.

Before Build from source, clean your C:\Users\...\.nuget\packages for remove all BepInEx/Harmony/Other entities, dont use BepInEx nuget.

Bug inside BeaverBuddies.Plugin.StartMod when calling HarmonyLib.Harmony.PatchAll :

> v0.6.3.1-b3de1fc-xsw
> InvalidOperationException: Incorrect code generation for exception block.
> System.Reflection.Emit.ILGenerator.EndExceptionBlock () (at <21f453231f9b419fb92653c858b2eaf6>:0)
> MonoMod.Utils._DMDEmit.Generate (MonoMod.Utils.DynamicMethodDefinition dmd, System.Reflection.MethodBase _mb, System.Reflection.Emit.ILGenerator il) (at <ce62e8a8bd7e458f939d11b9ac84886e>:0)
> MonoMod.Utils.DMDEmitMethodBuilderGenerator.GenerateMethodBuilder (MonoMod.Utils.DynamicMethodDefinition dmd, System.Reflection.Emit.TypeBuilder typeBuilder) (at <ce62e8a8bd7e458f939d11b9ac84886e>:0)
> MonoMod.Utils.DMDEmitMethodBuilderGenerator.GenerateCore (MonoMod.Utils.DynamicMethodDefinition dmd, System.Object context) (at <ce62e8a8bd7e458f939d11b9ac84886e>:0)
> MonoMod.Utils.DMDGenerator`1[TSelf].Generate (MonoMod.Utils.DynamicMethodDefinition dmd, System.Object context) (at <ce62e8a8bd7e458f939d11b9ac84886e>:0)
> MonoMod.Utils.DynamicMethodDefinition.Generate (System.Object context) (at <ce62e8a8bd7e458f939d11b9ac84886e>:0)
> MonoMod.Utils.DynamicMethodDefinition.Generate () (at <ce62e8a8bd7e458f939d11b9ac84886e>:0)
> HarmonyLib.MethodPatcher.CreateReplacement (System.Collections.Generic.Dictionary`2[System.Int32,HarmonyLib.CodeInstruction]& finalInstructions) (at <ce62e8a8bd7e458f939d11b9ac84886e>:0)
> HarmonyLib.PatchFunctions.UpdateWrapper (System.Reflection.MethodBase original, HarmonyLib.PatchInfo patchInfo) (at <ce62e8a8bd7e458f939d11b9ac84886e>:0)
> HarmonyLib.PatchClassProcessor.ProcessPatchJob (HarmonyLib.PatchJobs`1+Job[T] job) (at <ce62e8a8bd7e458f939d11b9ac84886e>:0)
> Rethrow as HarmonyException: Patching exception in method System.Void Timberborn.GameSaveRuntimeSystem.GameSaver::Save(Timberborn.GameSaveRuntimeSystem.QueuedSave queuedSave)
> HarmonyLib.PatchClassProcessor.ReportException (System.Exception exception, System.Reflection.MethodBase original) (at <ce62e8a8bd7e458f939d11b9ac84886e>:0)
> HarmonyLib.PatchClassProcessor.Patch () (at <ce62e8a8bd7e458f939d11b9ac84886e>:0)
> HarmonyLib.Harmony.<PatchAll>b__10_0 (System.Type type) (at <ce62e8a8bd7e458f939d11b9ac84886e>:0)
> HarmonyLib.CollectionExtensions.Do[T] (System.Collections.Generic.IEnumerable`1[T] sequence, System.Action`1[T] action) (at <ce62e8a8bd7e458f939d11b9ac84886e>:0)
> HarmonyLib.Harmony.PatchAll (System.Reflection.Assembly assembly) (at <ce62e8a8bd7e458f939d11b9ac84886e>:0)
> HarmonyLib.Harmony.PatchAll () (at <ce62e8a8bd7e458f939d11b9ac84886e>:0)
> BeaverBuddies.Plugin.StartMod () (at <fc23be3e00cd437f920991b330f6c4e5>:0)
> Timberborn.ModManagerScene.ModCodeStarter.StartMods () (at <789c05d7af744b0b95ae79c83b547ce1>:0)
> Timberborn.ModManagerScene.ModCodeStarter.Start () (at <789c05d7af744b0b95ae79c83b547ce1>:0)
> Timberborn.ModManagerSceneUI.ModManagerScenePanel.LoadModsAndStartGame () (at <60a1bcdd955e45f1b6f5ba8f3a7b23ed>:0)
> Timberborn.ModManagerSceneUI.ModManagerScenePanel.<InitializeModManagerPanel>b__13_0 (UnityEngine.UIElements.ClickEvent _) (at <60a1bcdd955e45f1b6f5ba8f3a7b23ed>:0)
> UnityEngine.UIElements.EventCallbackFunctor`1[TEventType].Invoke (UnityEngine.UIElements.EventBase evt) (at <40aa036fdcfe4396873350accf8ff004>:0)
> UnityEngine.UIElements.EventCallbackRegistry+DynamicCallbackList.Invoke (UnityEngine.UIElements.EventBase evt, UnityEngine.UIElements.BaseVisualElementPanel panel, UnityEngine.UIElements.VisualElement target) (at <40aa036fdcfe4396873350accf8ff004>:0)
> UnityEngine.UIElements.EventDispatchUtilities.HandleEventAcrossPropagationPath (UnityEngine.UIElements.EventBase evt, UnityEngine.UIElements.BaseVisualElementPanel panel, UnityEngine.UIElements.VisualElement target, System.Boolean isCapturingTarget) (at <40aa036fdcfe4396873350accf8ff004>:0)
> UnityEngine.UIElements.EventDispatchUtilities.DispatchToCachedElementUnderPointerOrPanelRoot (UnityEngine.UIElements.EventBase evt, UnityEngine.UIElements.BaseVisualElementPanel panel, System.Int32 pointerId, UnityEngine.Vector2 position) (at <40aa036fdcfe4396873350accf8ff004>:0)
> UnityEngine.UIElements.EventDispatchUtilities.DispatchToCapturingElementOrElementUnderPointer (UnityEngine.UIElements.EventBase evt, UnityEngine.UIElements.BaseVisualElementPanel panel, System.Int32 pointerId, UnityEngine.Vector2 position) (at <40aa036fdcfe4396873350accf8ff004>:0)
> UnityEngine.UIElements.PointerEventBase`1[T].Dispatch (UnityEngine.UIElements.BaseVisualElementPanel panel) (at <40aa036fdcfe4396873350accf8ff004>:0)
> UnityEngine.UIElements.EventDispatcher.ProcessEvent (UnityEngine.UIElements.EventBase evt, UnityEngine.UIElements.BaseVisualElementPanel panel) (at <40aa036fdcfe4396873350accf8ff004>:0)
> UnityEngine.UIElements.EventDispatcher.ProcessEventQueue () (at <40aa036fdcfe4396873350accf8ff004>:0)
> UnityEngine.UIElements.EventDispatcher.OpenGate () (at <40aa036fdcfe4396873350accf8ff004>:0)
> UnityEngine.UIElements.EventDispatcherGate.Dispose () (at <40aa036fdcfe4396873350accf8ff004>:0)
> UnityEngine.UIElements.EventDispatcher.ProcessEvent (UnityEngine.UIElements.EventBase evt, UnityEngine.UIElements.BaseVisualElementPanel panel) (at <40aa036fdcfe4396873350accf8ff004>:0)
> UnityEngine.UIElements.EventDispatcher.Dispatch (UnityEngine.UIElements.EventBase evt, UnityEngine.UIElements.BaseVisualElementPanel panel, UnityEngine.UIElements.DispatchMode dispatchMode) (at <40aa036fdcfe4396873350accf8ff004>:0)
> UnityEngine.UIElements.BaseVisualElementPanel.SendEvent (UnityEngine.UIElements.EventBase e, UnityEngine.UIElements.DispatchMode dispatchMode) (at <40aa036fdcfe4396873350accf8ff004>:0)
> UnityEngine.UIElements.PanelEventHandler.SendEvent (UnityEngine.UIElements.EventBase e, UnityEngine.EventSystems.BaseEventData sourceEventData) (at <80989b9abe4c41e3a14ffc8656da333e>:0)
> UnityEngine.UIElements.PanelEventHandler.OnPointerUp (UnityEngine.EventSystems.PointerEventData eventData) (at <80989b9abe4c41e3a14ffc8656da333e>:0)
> UnityEngine.EventSystems.ExecuteEvents.Execute (UnityEngine.EventSystems.IPointerUpHandler handler, UnityEngine.EventSystems.BaseEventData eventData) (at <80989b9abe4c41e3a14ffc8656da333e>:0)
> UnityEngine.EventSystems.ExecuteEvents.Execute[T] (UnityEngine.GameObject target, UnityEngine.EventSystems.BaseEventData eventData, UnityEngine.EventSystems.ExecuteEvents+EventFunction`1[T1] functor) (at <80989b9abe4c41e3a14ffc8656da333e>:0)
> UnityEngine.EventSystems.EventSystem:Update()
> 
>

**Comment by thomaswp** [comment]
Yes, thanks for posting! I'm aware of it. Trying to track down the
incompatibility, but may take some time.

**Comment by Cilyena** [comment]
I've identified the lines that crash. I put them in comments. 
File DeterminismService.cs :

    [HarmonyPatch(typeof(GameSaver), nameof(GameSaver.Save), typeof(QueuedSave))]
    public class GameSaverSavePatcher
    {
        public static bool IsSaving { get; set; }

        /*
        static bool Prefix(GameSaver __instance, QueuedSave queuedSave)
        {
            if (IsSaving || EventIO.IsNull) return true;
            TickingService ts = GetSingleton<TickingService>();
            if (ts == null) return true;
            ts.FinishFullTickAndThen(() =>
            {
                IsSaving = true;
                __instance.Save(queuedSave);
                IsSaving = false;
            });
            return false;
        */
    }

    [HarmonyPatch(typeof(Time), nameof(Time.time), MethodType.Getter)]
    public class TimeTimePatcher
    {
        private static float time = 0;

        public static void SetTicksSinceLoaded(int ticks)
        {
            time = ticks * Time.fixedDeltaTime;
        }
        /*
        static bool Prefix(ref float __result)
        {
            if (EventIO.IsNull) return true;
            __result = time;
            return false;
        }
        */
    }

With this change, the game start with Harmony (no HarmonyX).

**Comment by Cilyena** [comment]
Class TimeTimePatcher :
We have this problem : https://github.com/BepInEx/HarmonyX/wiki/Difference-between-Harmony-and-HarmonyX#patching-extern-methods-as-managed

**Comment by thomaswp** [comment]
Thanks so much for tracking this down - that was going to be my next step. 
* Patching Time.time is pretty important to the mod... I'll look into a workaround in Harmony, or possibly a more complex workaround by patching all methods that use it for game logic in Timberborn.
* I can also look into shipping HarmonyX, but that's not ideal.

I'm away for the rest of the week so it may be a while before I get to it.

**Comment by Cilyena** [comment]
>     * I can also look into shipping HarmonyX, but that's not ideal.

If HarmonyX is better than Harmony, and if HarmonyX has not been discontinued or abandoned by its developers, use it.

**Comment by Cilyena** [comment]
> If HarmonyX is better than Harmony, and if HarmonyX has not been discontinued or abandoned by its developers, use it.

It's done :
https://github.com/thomaswp/BeaverBuddies/blob/874cbf024bb0127ce1a53cf4f1315d178d00e0ec/BeaverBuddies/BeaverBuddies.csproj

I included HarmonyX in the final directory.
But now a new version of the game has been released and ValidatingGameLoader.LoadGame no longer exists. One more thing to fix.

I hope the work I've done will help you correct all these new changes.
